lib/iris/_constraints.py

```diff
@@ -478,3 +481,68 @@ def _cube_func(self, cube):
 
     def __repr__(self):
         return 'AttributeConstraint(%r)' % self._attributes
+
+
+class NameConstraint(Constraint):
+    """
+    Provides a simple Cube name based :class:`Constraint`, which matches
+    against each of the names provided, which may be either standard name,
+    long name, NetCDF variable name and/or the STASH from the attributes
+    dictionary.
+
+    The name constraint will only succeed if all of the provided names match.
+
+    """
+    def __init__(self, standard_name='none', long_name='none',
+                 var_name='none', STASH='none'):
```

Comment:
The trouble is that this does introduce another way of constraining on a stash code, which could now be done either by `NameConstraint` or `AttributeConstraint`, which goes against the zen of Python

Reply:
Again, this is in alignment with the `name` method. As it stands, it's completely possible to perform a stash constraint at least four other ways that I know of, excluding this PR, without thinking about it to much. IMHO `iris` is so anti-zen here already, that really the whole interface needs a re-design.

That said, the `NameConstraint` gives explicit control back to the user, so it does have value, and it also rights some wrongs to boot. The question that really needs answering here is whether stash should be part of the `name` method, `names` property and `NameConstraint`?

I know what my answer is, but I'm keen to know what others think...